spec/lib/collective_proof_attempt_spec.rb

```diff
@@ -131,51 +223,147 @@
+          allow_any_instance_of(UriFetcher::PostCaller).
+            to receive(:call).
+            and_raise(UriFetcher::UnfollowableResponseError.
+              new(last_response: OpenStruct.new(code: '422',
+                                                body: "*** Error: #{msg}")))
+        end
+
+        it 'raise a hets error' do
+          expect { cpa.run }.to raise_error(Hets::Errors::HetsFileError, msg)
         end
       end
 
-      it "each theorem's state is done in the end" do
-        theorems.each do |theorem|
-          expect(theorem.reload.state).to eq('done')
+      context 'with hets returning invalid json' do
+        before do
+          allow_any_instance_of(Hets::ProveCaller).to receive(:call).
+            and_return(StringIO.new('not valid json'))
+        end
+
+        it 'raise a parser error' do
+          expect { cpa.run }.to raise_error(Hets::JSONParser::ParserError)
         end
       end
 
-      it "ontology_version's state is done in the end" do
-        expect(ontology_version.reload.state).to eq('done')
+      context 'error handling' do
+        let(:error_message) { 'some error message' }
+        before do
+          allow_any_instance_of(Ontology).to receive(:import_proof).
+            and_raise(Hets::Errors::HetsFileError, error_message)
+          begin
+            cpa.run
+          rescue Hets::Errors::HetsFileError
+          end
+        end
+
+        it "each proof_attempt's proof_status is OPN" do
+          proof_attempts.each do |proof_attempt|
+            expect(proof_attempt.reload.proof_status).to eq(status_open)
+          end
+        end
+
+        it "each theorem's proof_status is OPN" do
+          theorems.each do |theorem|
+            expect(theorem.reload.proof_status).to eq(status_open)
+          end
+        end
+
+        it "each proof_attempt's is failed in the end" do
+          proof_attempts.each do |proof_attempt|
+            expect(proof_attempt.reload.state).to eq('failed')
+          end
+        end
+
+        it "each theorem's state is failed in the end" do
+          theorems.each do |theorem|
+            expect(theorem.reload.state).to eq('failed')
+          end
+        end
+
+        it "ontology_version's state is failed in the end" do
+          expect(ontology_version.reload.state).to eq('failed')
+        end
+
+        it "each proof_attempt's last_error contains the error message" do
+          proof_attempts.each do |proof_attempt|
+            expect(proof_attempt.reload.last_error).to include(error_message)
+          end
+        end
+
+        it "each theorem's last_error contains the error message" do
+          theorems.each do |theorem|
+            expect(theorem.reload.last_error).to include(error_message)
+          end
+        end
+
+        it "ontology_version's last_error contains the error message" do
+          expect(ontology_version.reload.last_error).to include(error_message)
+        end
       end
     end
   end
```

Comment:
For the same reason as in #1298 I decided to have some code duplication in the specs. Again, this operates once on `ontology_version` and once on `theorem`. I can create a context for the identical parts of theorem and ontology_version, if you want me to do it.